Integrate basic OpenAssetIO methods

feltech opened the pull request:
Part of #15. 

Integrate OpenAssetIO methods enough that the existing test cases pass.

That is, read-only fully and partially assetised situations (subject to our imposed limitation that `locateableContent.location` resolves to an absolute path), assuming the BAL library is bootstrapped successfully and configured correctly.

No effort has been put into error cases or the publishing workflow.

feltech committed:
[Tests] Use a single BAL library across all tests

The USD resolver plugin is only initialised once for the whole suite of
tests and we have no way to poke BAL to force a reinitialision with a
different library.

So combine all per-test `bal_libary.json` into a single library.

Add missing env var fixtures in the resolver `pytest` tests.

Signed-off-by: David Feltell <david.feltell@foundry.com>

foundrytom commented on src/resolver.cpp:
```diff
@@ -120,7 +144,12 @@ std::string UsdOpenAssetIOResolver::_CreateIdentifierForNewAsset(
 ArResolvedPath UsdOpenAssetIOResolver::_Resolve(const std::string &assetPath) const {
-  auto result = ArDefaultResolver::_Resolve(assetPath);
+  ArResolvedPath result;
+  if (manager_->isEntityReferenceString(assetPath)) {
```
[aside] We're def going to need to put in the C++ based prefix optimisation here!

foundrytom commented on src/resolver.cpp:
```diff
@@ -165,7 +194,15 @@ ArAssetInfo UsdOpenAssetIOResolver::_GetAssetInfo(const std::string &assetPath,
+  ArTimestamp result;
+  if (manager_->isEntityReferenceString(assetPath)) {
+    // Deliberately use a valid fixed timestamp, to force caching.
+    // TODO(DF): We need a "modificationTimestamp" trait to query from
```
[fix] Can we phrase this in terms of "is this something that we want to be a thing in OpenAssetIO" as it needs quite careful thought.

feltech replied:
Rephrased in e6617249

foundrytom commented on src/resolver.cpp:
```diff
@@ -180,7 +217,8 @@ std::shared_ptr<ArAsset> UsdOpenAssetIOResolver::_OpenAsset(
 
-  return ArDefaultResolver::_OpenAsset(resolvedPath);
+  return ArDefaultResolver::_OpenAsset(
+      ArResolvedPath{locationForEntity(resolvedPath).value_or(resolvedPath)});
```
[fix] Can we make sure we add TODOs for the areas we know we def need to revisit.

foundrytom replied:
[fix] `ArResolvedPath` wrapping is redundant.

feltech replied:
Removed redundant `ArResolvedPath` wrapping in 44e5e35e

feltech replied:
Added various TODOs in aaa71036. Anything I missed?

foundrytom commented on src/resolver.cpp:
```diff
@@ -201,3 +239,41 @@ std::shared_ptr<ArWritableAsset> UsdOpenAssetIOResolver::_OpenAssetForWrite(
+    manager_->resolve(
+        {std::move(*maybeEntityReference)}, {kLocateableContentTraitId}, readContext_,
+        [&traitsData]([[maybe_unused]] std::size_t idx,
+                      const openassetio::TraitsDataPtr &traitsData_) {
```
[fix] Can we use a different var name for the lambda - as this is our convention for private members? Makes it hard to grok.

feltech replied:
Renamed in 1d20438e

foundrytom commented on src/resolver.cpp:
```diff
@@ -201,3 +239,41 @@ std::shared_ptr<ArWritableAsset> UsdOpenAssetIOResolver::_OpenAssetForWrite(
   return ArDefaultResolver::_OpenAssetForWrite(resolvedPath, writeMode);
 }
+
+std::optional<std::string> UsdOpenAssetIOResolver::locationForEntity(
```
[fix] Return type is `ArResolvedPath`

feltech replied:
Whoops, left over from a previous incarnation. The return type of the method is `std::string`, but it's silently implicitly converted.

feltech replied:
Removed redundant `ArResolvedPath` wrapping in 44e5e35e

foundrytom replied:
Looking at it again, is this the right way around? Given that were using `value_or` with an `ArResolvePath`? See continuation of this in the other comment too.

foundrytom commented on src/resolver.cpp:
```diff
@@ -88,6 +98,9 @@ UsdOpenAssetIOResolver::UsdOpenAssetIOResolver() {
 
+  readContext_ = openassetio::Context::make(openassetio::Context::Access::kRead,
+                                            openassetio::Context::Retention::kTransient);
+
```
[fix] Add todo to set Locale

feltech replied:
Added in 467129ab

foundrytom commented on src/resolver.cpp:
```diff
@@ -120,7 +120,8 @@ std::string UsdOpenAssetIOResolver::_CreateIdentifier(
   } else {
-    identifier = ArDefaultResolver::_CreateIdentifier(assetPath, anchorAssetPath);
+    identifier = ArDefaultResolver::_CreateIdentifier(
+        assetPath, ArResolvedPath{locationForEntity(anchorAssetPath).value_or(anchorAssetPath)});
```
[fix] Redundan]t wrappping in `ArResolvedPath`

feltech replied:
`_CreateIdentifier` takes `ArResolvedPath`s and cannot implicitly convert from the `std::string` returned by `locationForEntity`.

foundrytom replied:
Yeah - this was from before, when it was already wrapped.

foundrytom replied:
Continuing on from the other comment on this topic - this is also re-wrapping `anchorAssetPath` in the fallback case. I guess were going to remove this soon anyway, but would it make more sense to just return an Ar path from the lookup?

feltech replied:
Initially we did return an `ArResolvedPath`, as you noticed, but then we need to unwrap it again in `_GetExtension`, so we went with returning a `std::string` as a the lowest common denominator. We also discussed a possible template-based solution, where we could vary the return type, but decided to keep things simple.

feltech replied:
Implemented a template-based solution in a623629c - can drop that commit if undesired.

foundrytom replied:
noice!

foundrytom commented on tests/resources/integration_test_data/bal_library.json:
```diff
@@ -0,0 +1,48 @@
+                {
+                    "traits": {
+                        "openassetio-mediacreation:content.LocatableContent": {
+                            "location": "file:///${TEST_DATA_ROOT}/assetized_child_ref_non_assetized_grandchild/floor1.usd"
```
[fix] I think these have one too many slashes, as `$TEST_DATA_ROOT` is an abs path.

feltech replied:
Interesting, looks like the `ArDefaultResolver` is pretty tolerant of too many leading `/`s.

feltech replied:
Updated in 291686cb

foundrytom replied:
I think it might be a posix thing, as it works in a shell too.

foundrytom reviewed (changes requested):
Great to see this coming together!

elliotcmorris commented on src/resolver.cpp:
```diff
@@ -165,7 +267,14 @@ ArAssetInfo UsdOpenAssetIOResolver::_GetAssetInfo(const std::string &assetPath,
-  auto result = ArDefaultResolver::_GetModificationTimestamp(assetPath, resolvedPath);
+  ArTimestamp result;
+  if (manager_->isEntityReferenceString(assetPath)) {
+    // Deliberately use a valid fixed timestamp, to force caching.
+    // TODO(DF): Consider how best to handle this via OpenAssetIO.
```
[Consider][Minor] I know you addressed this comment already, at at the risk of bikeshedding, this still comes off as though we had a direct need to force caching, rather than it merely being an (unfortunate) side effect.

```suggestion
    // Deliberately use a valid fixed timestamp, which will force caching
    // until we implement a more considered method.
    // TODO(DF): Consider how best to handle this via OpenAssetIO.
```

elliotcmorris reviewed (approved):
Nothing much further from me, I'd seen most of this already anyhow. Great work.

feltech committed:
[Core] Integrate read-only OpenAssetIO workflow

Part of #15.

Integrate simple read-only OpenAssetIO workflow, such that all xfail
tests now pass.

This just addresses the happy path. Many TODO comments added for things
to address.

Signed-off-by: David Feltell <david.feltell@foundry.com>
